stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -494,16 +453,31 @@ SWIFT_CC(swift)
 static void swift_task_cancelImpl(AsyncTask *task) {
   SWIFT_TASK_DEBUG_LOG("cancel task = %p", task);
 
-  // withStatusRecordLock has some special behavior for
-  // LockContext::Cancellation; the function only gets called
-  // when they don't apply.
-  withStatusRecordLock(task, LockContext::Cancellation,
-                       [&](ActiveTaskStatus &status) {
-    assert(status.isCancelled());
+  auto oldStatus = task->_private().Status.load(std::memory_order_relaxed);
+  auto newStatus = oldStatus;
+  while (true) {
+    if (oldStatus.isCancelled()) {
+      return;
+    }
+
+    // Set cancelled bit even if oldStatus.isStatusRecordLocked()
+    newStatus = oldStatus.withCancelled();
 
-    // Carry out the cancellation operations associated with all
-    // the active records.
-    for (auto cur: status.records()) {
+    if (task->_private().Status.compare_exchange_weak(oldStatus, newStatus,
+            /*success*/ std::memory_order_relaxed,
+            /*failure*/ std::memory_order_relaxed)) {
+      break;
+    }
+  }
+  newStatus.traceStatusChanged(task);
+
+  // Technically, more children could be added to the task between when
+  // we've just set the cancelled bit and when we grab the status record
+  // lock to propagate it but updateNewChildWithParentAndGroupState()
+  // handles this for the childre since we synchronize with parent's
+  // active task status and will see the canclled bit.
```

Comment:
I think this comment is a little misleading.  It doesn't actually matter for our local purposes that records can be added between when we set the cancelled bit and when we acquire the lock, because in any event we're going to visit all the records that are in place when we acquire the lock.  `updateNewChild...` isn't important for *those* records, it's important for the records being concurrently added that we *don't* see when we acquire the lock, i.e. the records that from our perspective are added strictly after us.

Reply:
Excellent point. Will update.